### ISSUE TITLE
Self referencing FK leads to "column reference is ambiguous" exception

### ISSUE THREAD

**Issue by lwqcz** [metabase/metabase#8371]
Hi there,

after I upgrade to version 0.30.1 from 0.30.0 I've stumble upon exception when I need to use table with metadata configured like one column is referencing via FK to another column in the same table.

I hope that *.gif posted below will support my description.

![metabase_self_foregin_key_issue](https://user-images.githubusercontent.com/39486334/44399367-a2f49880-a547-11e8-9220-0daeadc1ea7a.gif)

- Your browser and the version: Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.183 Safari/537.36 Vivaldi/1.96.1147.52
- Your operating system: Win10
- Your databases: Postgres
- Metabase version: 0.30.1
- Metabase hosting environment: Linux/Ubuntu (*.jar file)
- Metabase internal database: H2

**Comment by todind** [comment]
Same issue with connection to SQL Server.

**Comment by Arrowofdarkness** [comment]
+1 also happening on SQL Server

com.microsoft.sqlserver.jdbc.SQLServerException: 
Ambiguous column name 'Company_Audit_RecID'.


Full logs:

Aug 21 14:39:46 WARN metabase.query-processor :: {:status :failed,
:class java.util.concurrent.ExecutionException,
:error "com.microsoft.sqlserver.jdbc.SQLServerException: Ambiguous column name 'Company_Audit_RecID'.",
:stacktrace
["driver.generic_sql.query_processor$cancellable_run_query.invokeStatic(query_processor.clj:485)"
"driver.generic_sql.query_processor$cancellable_run_query.invoke(query_processor.clj:477)"
"driver.generic_sql.query_processor$run_query.invokeStatic(query_processor.clj:502)"
"driver.generic_sql.query_processor$run_query.invoke(query_processor.clj:497)"
"driver.generic_sql.query_processor$do_with_auto_commit_disabled.invokeStatic(query_processor.clj:535)"
"driver.generic_sql.query_processor$do_with_auto_commit_disabled.invoke(query_processor.clj:524)"
"driver.generic_sql.query_processor$do_in_transaction$fn__56571.invoke(query_processor.clj:540)"
"driver.generic_sql.query_processor$do_in_transaction.invokeStatic(query_processor.clj:539)"
"driver.generic_sql.query_processor$do_in_transaction.invoke(query_processor.clj:538)"
"driver.generic_sql.query_processor$run_query_without_timezone.invokeStatic(query_processor.clj:553)"
"driver.generic_sql.query_processor$run_query_without_timezone.invoke(query_processor.clj:552)"
"driver.generic_sql.query_processor$execute_query$fn__56595.invoke(query_processor.clj:577)"
"driver.generic_sql.query_processor$do_with_try_catch.invokeStatic(query_processor.clj:519)"
"driver.generic_sql.query_processor$do_with_try_catch.invoke(query_processor.clj:514)"
"driver.generic_sql.query_processor$execute_query.invokeStatic(query_processor.clj:574)"
"driver.generic_sql.query_processor$execute_query.invoke(query_processor.clj:570)"
"driver$fn__28276$G__28083__28283.invoke(driver.clj:104)"
"query_processor$execute_query.invokeStatic(query_processor.clj:54)"
"query_processor$execute_query.invoke(query_processor.clj:48)"
"query_processor.middleware.mbql_to_native$mbql__GT_native$fn__33022.invoke(mbql_to_native.clj:30)"
"query_processor.middleware.annotate_and_sort$annotate_and_sort$fn__31336.invoke(annotate_and_sort.clj:42)"
"query_processor.middleware.limit$limit$fn__32977.invoke(limit.clj:15)"
"query_processor.middleware.cumulative_aggregations$cumulative_aggregation$fn__32827.invoke(cumulative_aggregations.clj:58)"
"query_processor.middleware.cumulative_aggregations$cumulative_aggregation$fn__32827.invoke(cumulative_aggregations.clj:58)"
"query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__36951.invoke(results_metadata.clj:51)"
"query_processor.middleware.format_rows$format_rows$fn__32967.invoke(format_rows.clj:26)"
"query_processor.middleware.binning$update_binning_strategy$fn__31430.invoke(binning.clj:165)"
"query_processor.middleware.resolve$resolve_middleware$fn__30924.invoke(resolve.clj:483)"
"query_processor.middleware.expand$expand_middleware$fn__32707.invoke(expand.clj:607)"
"query_processor.middleware.add_row_count_and_status$add_row_count_and_status$fn__31010.invoke(add_row_count_and_status.clj:15)"
"query_processor.middleware.driver_specific$process_query_in_context$fn__32847.invoke(driver_specific.clj:12)"
"query_processor.middleware.resolve_driver$resolve_driver$fn__35494.invoke(resolve_driver.clj:15)"
"query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__31344$fn__31345.invoke(bind_effective_timezone.clj:9)"
"util.date$call_with_effective_timezone.invokeStatic(date.clj:82)"
"util.date$call_with_effective_timezone.invoke(date.clj:71)"
"query_processor.middleware.bind_effective_timezone$bind_effective_timezone$fn__31344.invoke(bind_effective_timezone.clj:8)"
"query_processor.middleware.cache$maybe_return_cached_results$fn__31519.invoke(cache.clj:149)"
"query_processor.middleware.catch_exceptions$catch_exceptions$fn__32758.invoke(catch_exceptions.clj:58)"
"query_processor$process_query.invokeStatic(query_processor.clj:135)"
"query_processor$process_query.invoke(query_processor.clj:131)"
"query_processor$run_and_save_query_BANG_.invokeStatic(query_processor.clj:249)"
"query_processor$run_and_save_query_BANG_.invoke(query_processor.clj:243)"
"query_processor$fn__36989$process_query_and_save_execution_BANG___36994$fn__36995.invoke(query_processor.clj:289)"
"query_processor$fn__36989$process_query_and_save_execution_BANG___36994.invoke(query_processor.clj:275)"
"query_processor$fn__37013$process_query_and_save_with_max_BANG___37018$fn__37019.invoke(query_processor.clj:310)"
"query_processor$fn__37013$process_query_and_save_with_max_BANG___37018.invoke(query_processor.clj:306)"
"api.dataset$fn__43466$fn__43469.invoke(dataset.clj:45)"
"api.common$fn__20623$invoke_thunk_with_keepalive__20628$fn__20629$fn__20630.invoke(common.clj:433)"],
:query
{:query {:source_table 1429, :filter ["time-interval" ["fk->" 17715 15596] -7 "day" {}]},
:type "query",
:parameters [],
:constraints {:max-results 10000, :max-results-bare-rows 2000},
:info
{:executed-by 1,
:context :ad-hoc,
:card-id nil,
:nested? false,
:query-hash [-102, 117, -24, -10, 9, 122, -101, -122, -49, -72, 80, 40, 51, -21, -113, 57, 127, 44, 74, 16, 75, -79, 81, -110, -79, 89, 56, -106, 127, -103, -62, -41],
:query-type "MBQL"}},
:expanded-query nil}

Aug 21 14:39:46 WARN metabase.query-processor :: Query failure: com.microsoft.sqlserver.jdbc.SQLServerException: Ambiguous column name 'Company_Audit_RecID'.
["query_processor$assert_query_status_successful.invokeStatic(query_processor.clj:217)"
"query_processor$assert_query_status_successful.invoke(query_processor.clj:210)"
"query_processor$run_and_save_query_BANG_.invokeStatic(query_processor.clj:250)"
"query_processor$run_and_save_query_BANG_.invoke(query_processor.clj:243)"
"query_processor$fn__36989$process_query_and_save_execution_BANG___36994$fn__36995.invoke(query_processor.clj:289)"
"query_processor$fn__36989$process_query_and_save_execution_BANG___36994.invoke(query_processor.clj:275)"
"query_processor$fn__37013$process_query_and_save_with_max_BANG___37018$fn__37019.invoke(query_processor.clj:310)"
"query_processor$fn__37013$process_query_and_save_with_max_BANG___37018.invoke(query_processor.clj:306)"
"api.dataset$fn__43466$fn__43469.invoke(dataset.clj:45)"
"api.common$fn__20623$invoke_thunk_with_keepalive__20628$fn__20629$fn__20630.invoke(common.clj:433)"]

Aug 21 14:39:46 DEBUG metabase.middleware :: POST /api/dataset 200 (244 ms) (15 DB calls). Jetty threads: 8/50 (11 busy, 1 idle, 0 queued)

**Comment by Arrowofdarkness** [comment]
DIdn't have a problem with it on 0.30!

**Comment by smerchek** [comment]
Also can confirm this same issue. Works in `v0.30.0` but not `v0.30.1` so I've reverted to `v0.30.0` for now.

For me, I wasn't referencing the same table. But we use the same column name for both the referencing table and the referenced table. e.g. `some_table.customer_id -> customer_table.customer_id`

**Comment by Arrowofdarkness** [comment]
I had to revert too. For me it was compliaining when I used *any* filters off a referenced table because it complained it was ambiguous. Don't apply a filter and it didn't care. 

Go directly to the referenced table and apply a filter there and it doesn't throw an error. When I reverted back to 0.30 it works both ways as is intended.

**Comment by lwqcz** [comment]
I do not know, I am not developer of Metabase, but this issue looks like quite serious or not? What about a fix ? @mazameli @salsakran and others ...

BTW: Was this change mentioned in 0.30.1 - change-log?

**Comment by dfx413** [comment]
Same problem here. 
When na name of the column in the referencing table is the same as the name of the referenced primary kay in another table I get this issue too.
This was working fine in 0.30.0. 
Had to revert back from 0.30.1

**Comment by wmorin** [comment]
Same issue here. We use a lot of fact tables referencing many dimensions primary keys.

For exemple:
- fact_order
      action_at
      **customer_id** (Foreign key to dim_customer->customer_id)
      amount
      currency

- dim_customer
  **customer_id**
  customer_name

Then we get this error:
  [Amazon](500310) Invalid operation: column reference "customer_id" is ambiguous

**Comment by guicouto** [comment]
Same problem here using 0.30.1

Log:
```
Query failure: org.postgresql.util.PSQLException: ERROR: column reference "id_servico_status" is ambiguous
  Position: 277
["query_processor$assert_query_status_successful.invokeStatic(query_processor.clj:217)"
 "query_processor$assert_query_status_successful.invoke(query_processor.clj:210)"
 "query_processor$run_and_save_query_BANG_.invokeStatic(query_processor.clj:250)"
 "query_processor$run_and_save_query_BANG_.invoke(query_processor.clj:243)"
 "query_processor$fn__36989$process_query_and_save_execution_BANG___36994$fn__36995.invoke(query_processor.clj:289)"
 "query_processor$fn__36989$process_query_and_save_execution_BANG___36994.invoke(query_processor.clj:275)"
 "api.card$run_query_for_card.invokeStatic(card.clj:582)"
 "api.card$run_query_for_card.doInvoke(card.clj:568)"
 "api.card$fn__46200$fn__46203.invoke(card.clj:589)"
 "api.card$fn__46200.invokeStatic(card.clj:588)"
 "api.card$fn__46200.invoke(card.clj:584)"
 "middleware$enforce_authentication$fn__55250.invoke(middleware.clj:116)"
 "api.routes$fn__55390.invokeStatic(routes.clj:61)"
 "api.routes$fn__55390.invoke(routes.clj:61)"
 "routes$fn__55477$fn__55478.doInvoke(routes.clj:108)"
 "routes$fn__55477.invokeStatic(routes.clj:103)"
 "routes$fn__55477.invoke(routes.clj:103)"
 "middleware$catch_api_exceptions$fn__55379.invoke(middleware.clj:424)"
 "middleware$log_api_call$fn__55356$fn__55358.invoke(middleware.clj:351)"
 "middleware$log_api_call$fn__55356.invoke(middleware.clj:350)"
 "middleware$add_security_headers$fn__55304.invoke(middleware.clj:253)"
 "core$wrap_streamed_json_response$fn__60666.invoke(core.clj:67)"
 "middleware$bind_current_user$fn__55255.invoke(middleware.clj:140)"
 "middleware$maybe_set_site_url$fn__55308.invoke(middleware.clj:277)"]
```

**Comment by wmorin** [comment]
Some knows a workaround to overpass this issue?

**Comment by mfpinhal** [comment]
> Some knows a workaround to overpass this issue?

Temporarily, you may convert the question to custom ad-hoc query (unfortunately, either by writing the query yourself from scratch or from logs), or just revert Metabase to the previous version as  https://github.com/metabase/metabase/issues/8371#issuecomment-414983081 did.

**Comment by leorochael** [comment]
Seems like a bad regression, but one that shouldn't be too hard for a novice to fix.

How one should go about trying to fix this issue?

**Comment by zedidas** [comment]
Same issue here. We upgraded from 0.29.3 to 0.30.1.

**Comment by someonegreat** [comment]
Same here. Anyone identify the issue yet?

**Comment by baires** [comment]
It was working before now on v0.30.1 I get the same error

**Comment by tschannerl** [comment]
same problem found, I went back to 0.30 is everything works. However I need this version 0.30.1 because there is language in Portuguese, help us ...

**Comment by o-rayer** [comment]
Same here. Any update on the fix ?

Thanks

**Comment by fetanchaud** [comment]
Hi all, problem is present in last 0.30.2 version.

**Comment by dfx413** [comment]
Not surprisingly, since it says "Open" :-) 
Would be nice to have it fixed, though.

**Comment by fetanchaud** [comment]
You are totally right, I just wanted to put the emphasis on that problem,
but I cannot help, I have no closure skills.

**Comment by dfx413** [comment]
Would like to help too but I also have no Clojure skills. Doesn't look like one can learn it in a week...

**Comment by MarkRx** [comment]
Blocker for using 0.30.x

**Comment by cgopalan** [comment]
Is this happening only for Postgres and SQL server? I could not simulate this error with H2 or MySQL.

**Comment by fetanchaud** [comment]
This happens with MySQL since v0.30.1,
we realized that the fact table name or alias is not inserted in the generated query :

below the fact_table foreign key id_dim is not prefixed with fact_table :

select 
  *
from 
  fact_table 
  left join dim_table as dim on (*fact_table*.id_dim = dim.id_dim)

**Comment by cgopalan** [comment]
I was able to successfully simulate the error in both mysql and oracle. The error occurs when I apply a filter in the UI question and do a self-join for the table to access the columns.

**Comment by senior** [comment]
This issue was caused by my changes in https://github.com/metabase/metabase/pull/8320. I know what the issue is and will have a fix ready tomorrow.

**Comment by MarkRx** [comment]
Will the fix be in 0.30.3?

**Comment by senior** [comment]
I just merged a fix (thanks @cgopalan) for this issue in the `release-0.30.3` branch. For those that hit this issue, it would be good to confirm that you no longer experience the issue on that branch.

@MarkRx It will be included in 0.30.3 which should be soon

**Comment by salsakran** [comment]
fixed in #8509
